cpp/src/arrow/python/udf.cc

```diff
@@ -0,0 +1,162 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/python/udf.h"
+
+#include <cstddef>
+#include <memory>
+#include <sstream>
+
+#include "arrow/compute/function.h"
+#include "arrow/python/common.h"
+
+namespace arrow {
+
+namespace py {
+
+Status VerifyArrayInput(const compute::ExecBatch& batch) {
+  for (auto value : batch.values) {
+    if (!value.is_array()) {
+      return Status::Invalid("Expected array input, but got ", value.type());
+    }
+  }
+  return Status::OK();
+}
+
+Status VerifyScalarInput(const compute::ExecBatch& batch) {
+  for (auto value : batch.values) {
+    if (!value.is_scalar()) {
+      return Status::Invalid("Expected scalar input, but got ", value.type());
+    }
+  }
+  return Status::OK();
+}
+
+Status VerifyArityAndInput(compute::Arity arity, const compute::ExecBatch& batch) {
+  if (!arity.is_varargs) {
+    bool match = static_cast<uint64_t>(arity.num_args) == batch.values.size();
+    if (!match) {
+      return Status::Invalid(
+          "Function Arity and Input data shape doesn't match, expected ", arity.num_args,
+          ", got ", batch.values.size());
+    }
+  } else {
+    bool match = static_cast<uint64_t>(arity.num_args) <= batch.values.size();
+    if (!match) {
+      return Status::Invalid("Required minimum number of arguments", arity.num_args,
+                             " in VarArgs function is not met.", ", Received ",
+                             batch.values.size());
+    }
+  }
+  return Status::OK();
+}
+
+Status ExecFunctionScalar(const compute::ExecBatch& batch, PyObject* function,
+                          const compute::Arity& arity, Datum* out) {
+  // num_args for arity varargs is arity.num_args, and for other arities,
+  // it is equal to the number of values in the batch
+  int64_t num_args =
+      arity.is_varargs ? static_cast<int64_t>(batch.values.size()) : arity.num_args;
+  PyObject* arg_tuple = PyTuple_New(num_args);
+  for (int arg_id = 0; arg_id < num_args; arg_id++) {
+    if (!batch[arg_id].is_scalar()) {
+      return Status::Invalid("Input type and data type doesn't match");
+    }
+    auto c_data = batch[arg_id].scalar();
```

Comment:
Note that when we want to handle functions that take only scalars, we will want to pass through `batch.length` eventually or else information will be lost. We may want to account for that now to avoid breaking changes.

Reply:
@lidavidm could you elaborate this a bit? I am not clear on this.

Reply:
Scalars do not represent a single value, they represent a column whose rows all have the same value. `batch.length` tells you the column length. This information should be reflected somehow.

Concretely: if you have a partition `a=2` and someone asks for `sum(a)`, you would expect a proper sum, not just `2`. But without `batch.length`, you have no way to compute the proper sum.

Reply:
Here we can explcityly pass an argument to the UDF such that users must add a parameter to get that information. 


```
def some_scalar_function(batch_length, *other_params):
      # do something
```

What do you think @lidavidm?

Reply:
If we need it, we should add it now to avoid breaking the API.

Reply:
CC @westonpace. It really only matters for functions which take only scalars, but it might be reasonable to add the argument for consistency. We may also want to put it inside some sort of context argument? So we can pass through things like the memory pool.

Reply:
@lidavidm if we need to pass in information like that, context will be the most suitable. 
Should we add it in a follow up PR or to this PR itself?

Reply:
It will be easier to add it now instead of breaking the API later

Reply:
@lidavidm 

```c++
class ARROW_PYTHON_EXPORT UdfContext {
  public:
    UdfContext(MemoryPool* pool) : pool_(pool) {}

    void SetConfig(const std::string &key, const std::string &value);

    std::string GetConfig(const std::string &key, const std::string &def = "");

    MemoryPool* pool() const { return pool_; }

  private:
    MemoryPool* pool_;
    std::unordered_map<std::string, std::string> config{};
};
```

How about an interface like this?

Reply:
Why not wrap KernelContext? And what is Set/GetConfig for?

Reply:
So does the context need to be static? Aren't we going to support holding a status which may be useful for stateful udfs?

Reply:
Ah, KernelContext has it all.

Reply:
What did you mean by holding a status?

Reply:
I think this is a good idea.  The important thing, from the python perspective, is that we pass in some kind of object.  It isn't as important I think that we get the API for this object exactly correct the first time around.  For example:

```
# Example UDF that takes zero arguments.
def random(context):
      batch_len = context.batch_length
      rand_numbers = get_random_numbers(batch_len)
      return rand_numbers
```

To start with we at least want to expose batch length.  That may be enough.  I think we want to add as few things as possible into the context object because these will be things we have to maintain.  The memory pool seems reasonable.  Things like cpu_info and use_threads I think we can add at a later date.

Reply:
Also, keep in mind, the batch length doesn't come from `KernelContext` but from the `ExecBatch` itself.  So this isn't `exec_chunksize`.

Reply:
Indeed, passing a context object as first argument is a good idea.

Reply:
Yes, I agree this is a nice addition. I have a question regarding the exposure of internals to Python. 
Should we expose the `KernelContext` to Python users? For instance getting the `MemoryPool` can be important. 
And also getting the `KernelState`.

Reply:
For instance, as @westonpace suggested, we get the `batch_length` from  `ExecBatch` not the `KernelContext`, but if we want to get the `MemoryPool` we need to extract it from the `KernelContext`. 

We can create an assoicative relationship for this attribute within the `UdfContext` object or extend from `KernelContext`. What should be the best? At the moment I am thinking about an associative relationship rather than an extension.

Reply:
@pitrou @lidavidm @westonpace WDYT?

Reply:
Don't worry about `KernelState`.  That is more important once we start handling stateful UDFs (e.g. aggregates) and we will probably wrap it anyways.  MemoryPool is already a well defined concept in python but there aren't many things a python user can do with it.  This would mostly be useful if the python user was going to call other Arrow APIs.  For example:

```
def random(context, one, two):
  return pc.add(one, two, memory_pool=context.memory_pool)
```

Reply:
@vibhatha You can just create a simple C++ struct and wrap it in Cython, for example:
```c++
struct ScalarUdfContext {
  MemoryPool* pool;
  int64_t batch_length;
};
```

And in Cython:
```cython
cdef class ScalarUdfContext:
    cdef:
        CScalarUdfContext context
    
    @property
    def memory_pool(self):
        return box_memory_pool(self.context.pool)
```
(you will need to implement `box_memory_pool`...)

Reply:
I don't have a strong preference on extend vs. association and I don't think we need to get it exactly right to start with.  If I had to pick I would go with association.

Reply:
@westonpace @pitrou thanks a lot for these pointers. Since this is too early to plan the end API, I will stick to an struct as @pitrou suggested. :)

Reply:
I have a question regarding exposing `ScalarUdfContext` to C++ Cython API. Because in the C++ we have to wrap the C++ context object into a `PyObject` and then pass it to the Python callable from C++. 
So a `wrap_scalar_udf_context` like function is required. Existing wrappers like that are defined in the `pyarrow.h` and  uses the `pyarrow_unwrap_####` methods. Can we adopt the same pattern here?

Reply:
If you mean the functions like those in pyarrow_api.h, that's probably best, yeah. Otherwise you could possibly use a Cython-side callback the way Flight does.

Reply:
Ah, because we have a separate _compute.so…Is it not possible to make it work perhaps by adding a _compute.pxd or something? There may be a better approach, I'll let Antoine or someone more familiar with Cython/CPython's API chime in

A callback will need some refactoring.

Reply:
cc @pitrou @amol- @jorisvandenbossche WDYT? Is there a better way? Would appreciate your insights.

Reply:
Indeed, I think a callback would be clearer.
You could have something like this on the C++ side:
```c++
using ScalarUdfWrapperCallback = std::function<PyObject*(
    PyObject* user_function, const ScalarUdfContext& context, PyObject* inputs)>;

ARROW_PYTHON_EXPORT
Status RegisterScalarFunction(
    PyObject* user_function, ScalarUdfWrapperCallback wrapper,
    const ScalarUdfOptions& options);
```

and then in Cython:
```cython

cdef _scalar_udf_callback(user_function, const CScalarUdfContext& c_context, inputs):
    # Create Python context
    context = ScalarUdfContext(...)
    return user_function(context, *inputs)
```

Reply:
@pitrou this is very helpful. I have added this functionality.